### ISSUE TITLE
RK4/Jit problem in GlobCurrent data

### ISSUE THREAD

**Issue by erikvansebille** [OceanPARCELS/parcels#70]
Simple advection now works for scipy in the GlibCurrent data, but somewhat surprising the jit doesn't work at all. The velocities are way off, and particles simply get swept outside of the domain very quickly.

I have made a test script in the [parcels-examples/tree/globcurrent_support](https://github.com/OceanPARCELS/parcels-examples/tree/globcurrent_support) branch, but because we don't have travis enabled on the parcels-examples repository I had to do it in the `__main__`

So you'll see that python `examples/test_globcurrent.py` works for scipy, but not for jit

**Comment by mlange05** [comment]
Ok, I think there are a few things going on here:
* Examples are enabled with travis in `parcels`, but you need to pull in the changes from `parcels-examples` through a separate branch in the main repo (eg.  commit 4624e89).
* The `starttime`, which comes from the grid itself, is much greater than the `endtime`, and thus we go into an assumed backward-in-time mode. I'm not sure that this is intentional, but it highlights that our time arguments convention in `execute`, including the use of `deltas`, is not entirely clear yet (eg. should `delta` indicate `starttime + delta`) and needs more rigorous unit testing to capture all possible uses and misuses.
* None of the above explains why the JIT mode gives a different field sampling result than the Python interpolators on the same time frame, so this is definitely a bug and I'll keep digging...

**Comment by erikvansebille** [comment]
Thank you @mlange05 for taking a look
Yes, I saw the point about the starttime too, and agree it will need to be fixed. Another issue is that `time_origin` is not properly parsed because the time units are spelled `Unit` rather than `units`.

However, the bigger issue is the different velocities in JIT and scipy. printing e.g. `u4` at every step in the AdvectionRK4 shows that they are very different. I expect something with the unit conversion?

**Comment by mlange05** [comment]
Ok, I think I have this triaged: The `lat` and `lon` arrays from the file are 64-bit doubles, but our JIT code assumes 32-bit floats. This causes all the grid indexing to go out of the window and hence erroneous field sampling. Since re-casting might be kind a bit tedious (and potentially slow) I would argue we should be able to adjust our JIT code to the data type in the file, but that might require a careful re-think of the engine. 

I'll try to come up with a clean fix later, but in the mean-time feel free to cast `lon` and `lat` to the right data type in the field constructor to bypass this for now.

**Comment by erikvansebille** [comment]
Thanks @mlange05, I can indeed confirm that adding the four lines
```
    grid.U.lon = np.float32(grid.U.lon)
    grid.U.lat = np.float32(grid.U.lat)
    grid.V.lon = np.float32(grid.V.lon)
    grid.V.lat = np.float32(grid.V.lat)
```
to `test_globcurrent_particles(mode)` fixes the JIT issue. Do you want to close this bug, or keep it open until a proper solution has been found?

**Comment by mlange05** [comment]
Ok, for now can we please add a sanity check to the field constructor that ensures that lat/lon are of the right type, similar to the one we do for the data itself (field.py, lines 117-119)? This is much safer than just fixing it in the test. We might also want to file a (low-priority) issue stating that we should at some point let the JIT engine adjust the types in thhe sampling code to the data provided in the files.

**Comment by mlange05** [comment]
Fixed in #71, so closing.
